advanced_source/cpp_cuda_graphs.rst

```diff
@@ -0,0 +1,177 @@
+
+The main training loop consists of the several steps and depicted in the
+following code chunk:
+
+.. code-block:: cpp
+
+  for (auto& batch : data_loader) {
+    auto data = batch.data.to(device);
+    auto targets = batch.target.to(device);
+    optimizer.zero_grad();
+    auto output = model.forward(data);
+    auto loss = torch::nll_loss(output, targets);
+    loss.backward();
+    optimizer.step();
+  }
+
+Which are basically a forward pass, backward pass and weight updates.
+
+In this tutorial we will be applying CUDA Graph on all the compute steps via whole-network
+graph capture. But before doing so, we need to slightly modify the source code. What we need
+to do is preallocate tensors for reusing them in the main training loop like so:
+
+.. code-block:: cpp
+
+  torch::TensorOptions FloatCUDA =
+      torch::TensorOptions(device).dtype(torch::kFloat);
+  torch::TensorOptions LongCUDA =
+      torch::TensorOptions(device).dtype(torch::kLong);
+
+  torch::Tensor data = torch::zeros({kTrainBatchSize, 1, 28, 28}, FloatCUDA);
+  torch::Tensor targets = torch::zeros({kTrainBatchSize}, LongCUDA);
+  torch::Tensor output = torch::zeros({1}, FloatCUDA);
+  torch::Tensor loss = torch::zeros({1}, FloatCUDA);
+
+  for (auto& batch : data_loader) {
+    data.copy_(batch.data);
+    targets.copy_(batch.target);
+    training_step(model, optimizer, data, targets, output, loss);
+  }
+
+Where ``training_step`` simply consists of forward and backward passes with corresponding optimizer calls:
+
+.. code-block:: cpp
+
+  void training_step(
+      Net& model,
+      torch::optim::Optimizer& optimizer,
+      torch::Tensor& data,
+      torch::Tensor& targets,
+      torch::Tensor& output,
+      torch::Tensor& loss) {
+    optimizer.zero_grad();
+    output = model.forward(data);
+    loss = torch::nll_loss(output, targets);
+    loss.backward();
+    optimizer.step();
+  }
+
+PyTorch’s CUDA Graphs API is relying on Stream Capture which in our case would used as following:
+
+.. code-block:: cpp
+
+  at::cuda::CUDAGraph graph;
+  at::cuda::CUDAStream captureStream = at::cuda::getStreamFromPool();
+  at::cuda::setCurrentCUDAStream(captureStream);
+
+  graph.capture_begin();
+  training_step(model, optimizer, data, targets, output, loss);
+  graph.capture_end();
+
+Before actual graph capture it is important to run several warm-up iterations on side stream to
+prepare CUDA cache as well as CUDA libraries (like CUBLAS and CUDNN) that will be used during
+the training:
+
+.. code-block:: cpp
+
+  at::cuda::CUDAStream warmupStream = at::cuda::getStreamFromPool();
+  at::cuda::setCurrentCUDAStream(warmupStream);
+  for (int iter = 0; iter < num_warmup_iters; iter++) {
+    training_step(model, optimizer, data, targets, output, loss);
+  }
+
+After successful graph capture we can replace ``training_step(model, optimizer, data, targets, output, loss);``
+call via ``graph.replay();`` do the training step.
+
+The full source code is available in GitHub.
+
```

Comment:
I feel like this can be a separate section. Maybe "Comparing  Training Results" or something like that?